#!/bin/bash
#SBATCH --output=./slurm-out/kalindx-%j.out
#SBATCH --job-name=make_kallisto_indices
#SBATCH --mem=128000
#SBATCH -c 2

#JOB LOG HEADER
perl -E 'say"="x80'; echo "JOB STARTED: `date`"; echo "NODE: `hostname`"; echo "SCRIPT ${0}:"; echo "JOB ID: ${SLURM_JOB_ID}"; cat $0; perl -E 'say"="x80'

#SOFTWARE REQUIREMENTS

#VARIABLES
BEDFILE=$(pwd)/rseqc_gene_models.bed
GENOME=$(pwd)/Mus_musculus.GRCm39.dna.primary_assembly.fa
INDEX_ID=EnsMm_${ENSEMBL_RELEASE}_Kallisto

echo ensembl release: $ENSEMBL_RELEASE
#COMMAND(s) TO RUN
# Build Kallisto Index for Exons using bed file
# generated by ensembl_make_rseqc_bedfile.sh

kallisto index -i ${INDEX_ID}_exon\
	 <(\
	   bedtools getfasta\
		    -split\
		    -nameOnly\
		    -fi $GENOME\
                    -bed $BEDFILE\
    )

# Build Kallisto Index for total transcript using bed file
# generated by ensembl_make_rseqc_bedfile.sh

kallisto index -i ${INDEX_ID}_total\
	 <(\
	   bedtools getfasta\
		    -nameOnly\
		    -fi $GENOME\
                    -bed $BEDFILE\
    )


#JOB LOG FOOTER
perl -E 'say"="x80'; echo "JOB COMPLETED: `date`"; perl -E 'say"="x80'
